### DISCUSSION TITLE
RLS with roles. Is it possible to stack policies?

### DISCUSSION THREAD

**Discussion by procrates** [supabase/supabase#7314] [Questions] [answered]
I have customized this example for my usecase: [https://github.com/supabase/supabase/tree/master/examples/slack-clone/nextjs-slack-clone](https://github.com/supabase/supabase/tree/master/examples/slack-clone/nextjs-slack-clone)

I have 3 roles, and a table that the user role can add, update and read their own posts.

Now i want an admin account to be able to do everything to those posts.

Anyone have any good solutions for this? 

I have tried to stack policies so that admins can do everything and users can access their own posts.

Thanks for any help I can get on this!

**Comment by procrates** [comment]
Anyone knows? I would like to avoid using service role at all costs since I am already using roles. 

Just need to be able to let users create their own stuff and be able to view delete and update their own stuff. But I also want an admin to be able to do the same thing.

The role check function is this. Btw, a user can have multiple roles also. But if any role match the permissions it should be able to do it.

```
declare
  bind_permissions int;
begin
  select count(*)
  from public.role_permissions
  inner join public.user_roles on role_permissions.role = user_roles.role
  where role_permissions.permission = authorize.requested_permission
    and user_roles.user_id = authorize.user_id
  into bind_permissions;

  return bind_permissions > 0;
end;
```

**Comment by GaryAustin1** [comment, marked as answer]
Not sure what you mean by "stack policies"
You can use and/or in your policy to check different things.  The policy is basically an SQL WHERE clause so you have a lot of flexibility to call functions or select from tables, etc. and check for basic user in one statement then and use OR to grant another user type access. 

This repository might help with ideas:  https://github.com/supabase-community/supabase-custom-claims
And a couple from admin role search here:
https://github.com/supabase/supabase/discussions/1138
https://github.com/supabase/supabase/discussions/7060

**Reply by procrates** [reply]
Great answer!

I will definitly try custom claims, just need to copy the database to dev since the app is in beta production.
The stack policies means:

Authorized are allowed to delete posts from other users
```
DELETE Allow authorized to delete based on user role authorize('posts.delete'::app_permission, uid())
```

Users are allowed to delete their own posts
```
DELETE Enable delete for users on posts based on user_id (uid() = user_id)
```

And this is the authorize function that gets userroles from user_roles table and permissions from the role_permissions table:
```
declare
  bind_permissions int;
begin
  select count(*)
  from public.role_permissions
  inner join public.user_roles on role_permissions.role = user_roles.role
  where role_permissions.permission = authorize.requested_permission
    and user_roles.user_id = authorize.user_id
  into bind_permissions;

  return bind_permissions > 0;
end;
```

**Reply by GaryAustin1** [reply]
Regarding "stacking" policies.  I normally just use an AND or OR statement in the same policy.  But mulitple policies on same thing are by default are OR'd.  You can force multiple policies to AND also.  
See about 1/3 down here:  https://www.postgresql.org/docs/current/ddl-rowsecurity.html
